File: ironapi/models/ratchetindex.js
const mongoose = require('mongoose');

const RatchetIndexSchema = new mongoose.Schema({
  user: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true },
  ratchetIndex: { type: String },  //index of remote user's public key
  ratchetValue: { type: String },  //encrypted message key
  crank: { type: String },  //nounce
  signature: { type: String },  ///XChaCha signature
  device: { type: String, },
  active: { type: Boolean, default: true },
  kdfNonce: { type: String},  //this nonce is used for encrypted backup secrets
  //these are not used for baseCircleObjects
  cipher: { type: String },  //user for all but CircleObjects
  cipherCrank: { type: String },  //nonce
  cipherSignature: { type: String },  //mac signature
  senderRatchetPublic: { type: String },  //sender's ratcheted public key

}, { timestamps: false }, { collection: 'ratchetindex' });

//deprecated
RatchetIndexSchema.statics.fromJson = function (json) {
  let local = this(json);

  if (local.kdfNonce == null) local.kdfNonce = undefined;
  if (local.cipher == null) local.cipher = undefined;
  if (local.cipherCrank == null) local.cipherCrank = undefined;
  if (local.cipherSignature == null) local.cipherSignature = undefined;
  if (local.senderRatchetPublic == null) local.senderRatchetPublic = undefined;

  return local;
}

RatchetIndexSchema.statics.new = function (json) {
  let local = this(json);

  if (local.kdfNonce == null) local.kdfNonce = undefined;
  if (local.cipher == null) local.cipher = undefined;
  if (local.cipherCrank == null) local.cipherCrank = undefined;
  if (local.cipherSignature == null) local.cipherSignature = undefined;
  if (local.senderRatchetPublic == null) local.senderRatchetPublic = undefined;

  return local;

}


mongoose.model('RatchetIndex', RatchetIndexSchema);
module.exports = mongoose.model('RatchetIndex');

File: ironclient/lib/models/ratchetindex.dart
class RatchetIndex {
  String ratchetIndex;
  String user;
  String ratchetValue;
  String userIndex;
  String crank;
  String signature;
  String? kdfNonce;
  String device;
  bool active;

  ///these are not used in base CircleObjects, only in Events (soon to be Lists and Recipes), Circle names, backgrounds, templates, etc
  String? cipher;
  String? cipherCrank;
  String? cipherSignature;
  String? senderRatchetPublic;

  RatchetIndex(
      {required this.ratchetIndex,
      required this.user,
      required this.crank,
      required this.signature,
      required this.ratchetValue,
      this.active = true,
      this.userIndex = '',
      this.device = '',
      this.kdfNonce,
      this.cipher,
      this.cipherCrank,
      this.cipherSignature,
      this.senderRatchetPublic});

  factory RatchetIndex.fromJson(Map<String, dynamic> json) => RatchetIndex(
        ratchetIndex: json['ratchetIndex'],
        user: json['user'],
        active: json['active'] ?? true,
        crank: json['crank'],
        device: json['device'] ?? '',
        signature: json['signature'],
        kdfNonce: json['kdfNonce'],
        ratchetValue: json['ratchetValue'],
        cipher: json['cipher'],
        cipherCrank: json['cipherCrank'],
        cipherSignature: json['cipherSignature'],
        senderRatchetPublic: json['senderRatchetPublic'],
        //userIndex: json['userIndex'],
      );

  factory RatchetIndex.blank() => RatchetIndex(
        ratchetIndex: '',
        user: '',
        active: true,
        crank: '',
        signature: '',
        ratchetValue: '',
      );

  Map<String, dynamic> toJson() => {
        'ratchetIndex': ratchetIndex,
        'crank': crank,
        'signature': signature,
        'user': user,
        'active': active,
        'device': device,
        'ratchetValue': ratchetValue,
        'kdfNonce': kdfNonce,
        'cipher': cipher,
        'cipherCrank': cipherCrank,
        'cipherSignature': cipherSignature,
        'senderRatchetPublic': senderRatchetPublic,
      };
}

class RatchetIndexCollection {
  final List<RatchetIndex> ratchetIndexes;

  RatchetIndexCollection.fromJSON(Map<String, dynamic> json, String key)
      : ratchetIndexes = (json[key] as List)
            .map((json) => RatchetIndex.fromJson(json))
            .toList();
}
